src/NzbDrone.Core/Notifications/Sendgrid/SendGridService.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Linq;
+using NLog;
+using SendGrid;
+using SendGrid.Helpers.Mail;
+using System.Net.Http.Headers;
+using FluentValidation.Results;
+using NzbDrone.Core.Notifications.Sendgrid.Models;
+
+namespace NzbDrone.Core.Notifications.Sendgrid
+{
+    public interface ISendGridService
+    {
+        SendGridResponse Send(SendgridSettings settings, string subject, string body);
+        ValidationFailure Test(SendgridSettings settings);
+    }
+    
+    
+    public class SendGridService : ISendGridService
+    {
+        private readonly Logger _logger;
+        private SendGridClient _client;
+        private static readonly string MESSAGE_ID_HEADER_NAME = "X-Message-Id";
+
+        public SendGridService(Logger logger)
+        {
+            _logger = logger;
+        }
+        
+        public SendGridResponse Send(SendgridSettings settings, string subject, string body)
+        {
+            _client = new SendGridClient(settings.ApiKey);
+            
+            var emailMessage = new SendGridMessage()
+            {
+                From = new EmailAddress(settings.From),
+                Subject = subject, 
+                HtmlContent = body
+            };
+            
+            emailMessage.AddTo(new EmailAddress(settings.To));
+            
+            return ProcessResponse(_client.SendEmailAsync(emailMessage).Result);
+        }
+
+        public ValidationFailure Test(SendgridSettings settings)
+        {
+            const string body = "Success! You have properly configured your SendGrid notification settings";
+
+            try
+            {
+                Send(settings, "Sonarr - Test Notification", body);
+            }
+            catch (Exception ex)
+            {
+                _logger.Error(ex, "Unable to send test email though SendGrid");
+                return new ValidationFailure("apiKey", "Unable to send test email though SendGrid");
+            }
+
+            return null;
+        }
+
+        private SendGridResponse ProcessResponse(Response response)
+        {
+            if (response.StatusCode.Equals(System.Net.HttpStatusCode.Accepted)
```

Comment:
Do we need to check that `response` is not null before accessing? Previously that was done in `ToMailResponse`, after being used here, but if it can be null we should verify it isn't here.

Reply:
I believe that it can be null, so check added there.